Enhance integration tests relating to transaction metadata and fees

jonathanknowles opened the pull request:
# Issue Number

ADP-1009

# Overview

This PR doesn't provide a solution for ADP-1009, but it does add integration test coverage relating to transaction metadata and fees.

In principle, the following two endpoints, if called on a wallet with the same starting UTxO and coin selection criteria, should return **_identical_** transaction fees (modulo random selection):

- [`selectCoins`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/selectCoins)
- [`postTransaction`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransaction)

# Modifications made by this PR

This PR:

- Adjusts `TRANSMETA_CREATE_01` so that it tests with a variety of different coin selection criteria:
    - no metadata;
    - simple textual metadata; 
    - the set of metadata recorded in ADP-1005.
- Adjusts `TRANSMETA_CREATE_01` so that it first performs a dry-run coin selection with the [`selectCoins`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/selectCoins) endpoint.
- Adjusts `TRANSMETA_CREATE_01` so that it cross-checks the fee obtained from calling [`selectCoins`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/selectCoins) with the fee obtained from calling [postTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransaction), failing if the fees are not identical.

This PR provides evidence that:
- that the [`selectCoins`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/selectCoins) and [`postTransaction`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransaction) endpoints **_do_** generate selections with **_identical_** fees, **_provided_** the coin selection criteria are identical.
- that we **_can_** indeed make a transaction that includes the metadata recorded in ADP-1005, and that the transaction is **_not_** rejected by the ledger for having a fee that is too low.

jonathanknowles committed:
Add transaction metadata for ADP-1005.

jonathanknowles committed:
Extract out `TRANSMETA_CREATE_01` into separate spec function.

This allows the test to be parameterized over different types of
transaction metadata.

jonathanknowles committed:
Add ADP-1005 metadata to `TRANSMETA_CREATE_01` spec.

jonathanknowles committed:
Add function `computeApiCoinSelectionFee`.

jonathanknowles committed:
Add explicit fee test to `TRANSMETA_CREATE_01` spec.

jonathanknowles committed:
Add explicit fee test to `HW_WALLETS_04: Can get coin selection`.

jonathanknowles committed:
Extract out function `spec_selectCoins` from `HW_WALLETS_04`.

This will allow it to be applied to different types of metadata.

jonathanknowles committed:
Move `txMetadata_ADP_1005` into `Framework.TestData`.

This allows it to be used from test cases in different modules.

jonathanknowles committed:
Make `spec_selectCoins` embed tx metadata in `selectCoin` requests.

jonathanknowles committed:
Add test cases with tx metadata to `HW_WALLETS_04`.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/HWWallets.hs:
```diff
@@ -375,7 +400,111 @@ spec = describe "SHELLEY_HW_WALLETS" $ do
+            let apiCoinSelection =
+                    getFromResponse Prelude.id coinSelectionResponse
+            let fee = computeApiCoinSelectionFee apiCoinSelection
+            fee `shouldBe` Coin expectedFee
```
We check here if `expectedFee` equals result of hand written `computeApiCoinSelectionFee`, but might it be that Ledger hadrware device calculates it wrong?

jonathanknowles replied:
> We check here if `expectedFee` equals result of hand written `computeApiCoinSelectionFee`, but might it be that Ledger hadrware device calculates it wrong?

Does the ledger hardware device perform its own fee computation, or adjust the ada quantities in any way?

I was under the impression that it just signs the the transaction provided to it.

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/HWWallets.hs:
```diff
@@ -375,7 +400,111 @@ spec = describe "SHELLEY_HW_WALLETS" $ do
+            let apiCoinSelection =
+                    getFromResponse Prelude.id coinSelectionResponse
+            let fee = computeApiCoinSelectionFee apiCoinSelection
+            fee `shouldBe` Coin expectedFee
```
Perhaps it would be good to also make a call to `postTransaction` in this test, to provide more direct confirmation that the fee is identical.

jonathanknowles replied:
Fixed by moving the dry-run call to `selectCoins` from `HWWallets` to `Transactions`.

jonathanknowles committed:
Move `computeApiCoinSelectionFee` to `Framework.DSL` module.

jonathanknowles committed:
Perform coin selection "dry-run" within "TRANSMETA_CREATE_01".

By doing this, we can confirm that the 'selectCoins' endpoint produces a
selection whose fee is identical to the selection  produced by the
'postTransaction' endpoint.

jonathanknowles committed:
Add "transaction without any metadata" case to `TRANSMETA_CREATE_01`.

This allows us to compare the fees before and after adding metadata.

jonathanknowles committed:
Introduce record type to clarify parameters of `TRANSMETA_CREATE_01`.

jonathanknowles committed:
Allow different numbers of outputs in `TRANSMETA_CREATE_01`.

jonathanknowles committed:
Add a test with the exact output quantities recorded in ADP-1005.

jonathanknowles committed:
Remove redundant test from `Test.Integration.Scenario.API.Shelley.HWWallets`.

Creating coin selections (with and without metadata) is now covered in:

`Test.Integration.Scenario.API.Shelley.Transactions`.

piotr-iohk reviewed (approved):
LGTM.
